File: Insight/src/agents/sentiment.ts
/**
 * Sentiment Agent (Tier 1)
 * Applies rule-based + lexicon sentiment scoring to each post.
 * Outputs a normalized score from -100 (most negative) to +100 (most positive).
 * 
 * For the prototype we use a lightweight lexicon-based approach so it works
 * entirely in the browser without requiring a model download.
 * The ml.worker.ts ONNX approach can be swapped in for production.
 */
import type { RedditPost, ProcessedPost } from '@/types';

const NEGATIVE_TERMS = [
  'terrible', 'broken', 'worst', 'useless', 'awful', 'hate', 'frustrated', 'angry',
  'disappointed', 'wrong', 'bad', 'fail', 'fails', 'failed', 'failure', 'disaster',
  'impossible', 'annoying', 'painful', 'nightmare', 'ridiculous', 'unacceptable',
  'confusing', 'slow', 'crash', 'bug', 'issue', 'problem', 'poor', 'lacking',
  'missing', 'broken', 'garbage', 'trash', 'waste', 'outrage', 'scam', 'locked',
  'hidden', 'gimmick', 'useless', 'unhelpful', 'overwhelming', 'exhausting',
];

const POSITIVE_TERMS = [
  'great', 'excellent', 'love', 'amazing', 'fantastic', 'perfect', 'wonderful',
  'helpful', 'easy', 'simple', 'intuitive', 'fast', 'reliable', 'solid',
  'recommend', 'impressed', 'delighted', 'smooth', 'clean', 'best',
];

const INTENSIFIERS = ['very', 'extremely', 'really', 'absolutely', 'completely', 'totally'];
const NEGATORS = ['not', 'no', 'never', 'nothing', 'neither', "don't", "doesn't", "can't", "won't"];

function tokenize(text: string): string[] {
  return text.toLowerCase().replace(/[^a-z0-9\s'-]/g, ' ').split(/\s+/).filter(Boolean);
}

function scoreLexicon(tokens: string[]): number {
  let score = 0;
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]!;
    const prevToken = tokens[i - 1] ?? '';
    const prev2Token = tokens[i - 2] ?? '';
    const isNegated = NEGATORS.includes(prevToken) || NEGATORS.includes(prev2Token);
    const isIntensified = INTENSIFIERS.includes(prevToken) ? 1.5 : 1;

    if (NEGATIVE_TERMS.includes(token)) {
      score += isNegated ? +15 : (-20 * isIntensified);
    } else if (POSITIVE_TERMS.includes(token)) {
      score += isNegated ? -15 : (20 * isIntensified);
    }
  }
  return Math.max(-100, Math.min(100, score));
}

function getWeekKey(timestamp: number): string {
  const d = new Date(timestamp * 1000);
  const jan1 = new Date(d.getFullYear(), 0, 1);
  const week = Math.ceil(((d.getTime() - jan1.getTime()) / 86400000 + jan1.getDay() + 1) / 7);
  return `${d.getFullYear()}-W${String(week).padStart(2, '0')}`;
}

export async function runSentimentAgent(
  posts: RedditPost[],
  onProgress: (msg: string) => void,
): Promise<ProcessedPost[]> {
  onProgress('Sentiment Agent: Scoring post sentiment...');
  await new Promise(r => setTimeout(r, 300));

  const processed: ProcessedPost[] = posts.map(post => {
    const fullText = `${post.title} ${post.selftext}`;
    const tokens = tokenize(fullText);
    let rawScore = scoreLexicon(tokens);

    // Boost negativity signal from upvote ratio (highly-upvoted complaints = strong signal)
    if (rawScore < 0 && post.upvote_ratio > 0.9) {
      rawScore = Math.max(-100, rawScore * 1.2);
    }

    const sentimentScore = Math.round(rawScore);
    const sentimentLabel: 'positive' | 'negative' | 'neutral' =
      sentimentScore > 15 ? 'positive' : sentimentScore < -15 ? 'negative' : 'neutral';

    return {
      ...post,
      cleanText: fullText.trim(),
      sentimentScore,
      sentimentLabel,
      weekKey: getWeekKey(post.created_utc),
    };
  });

  const negCount = processed.filter(p => p.sentimentLabel === 'negative').length;
  onProgress(`Sentiment Agent: Scored ${posts.length} posts (${negCount} negative signals) ✓`);
  return processed;
}

File: backend/agents/sentiment.py
"""
Sentiment Agent
Hybrid approach: VADER lexicon + rule-based amplification.
Outputs: positive / neutral / negative label + normalized score (-1.0 to +1.0)
Production extension: swap VADER for trained TF-IDF+SVC model.
"""
import time
import re
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from memory import SharedMemory, ProcessedPost, AgentLog

_analyzer = SentimentIntensityAnalyzer()

# Amplification terms: product-specific negative signals
STRONG_NEGATIVE = {
    "fraud", "scam", "broken", "crash", "fails", "failed", "failure",
    "worst", "terrible", "horrible", "useless", "refund", "cancel",
    "stolen", "missing", "lost", "damaged", "delay", "delayed", "late",
    "wrong", "incorrect", "never", "never received", "pathetic", "garbage",
    "rip off", "ripoff", "cheated", "cheating", "lies", "lying", "blocked",
    "unacceptable", "disappointed", "disgrace", "waste", "awful"
}

STRONG_POSITIVE = {
    "excellent", "amazing", "fantastic", "best", "love", "great",
    "superb", "perfect", "wonderful", "recommend", "impressed", "reliable",
    "fast", "smooth", "easy", "convenient", "helpful", "awesome"
}


def _clean_text(text: str) -> str:
    text = re.sub(r"https?://\S+", "", text)
    text = re.sub(r"[^\w\s.,!?'-]", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text[:1500]


def _score(title: str, body: str) -> tuple[str, float, float]:
    combined = f"{title}. {body}".lower().strip()
    clean = _clean_text(combined)

    # VADER base score
    scores = _analyzer.polarity_scores(clean)
    compound = scores["compound"]  # -1.0 to +1.0

    # Amplify with domain-specific terms
    words = set(clean.split())
    neg_hits = len(words & STRONG_NEGATIVE)
    pos_hits = len(words & STRONG_POSITIVE)

    compound -= neg_hits * 0.07
    compound += pos_hits * 0.05
    compound = max(-1.0, min(1.0, compound))

    # Label with stricter threshold for complaints (domain-tuned)
    if compound <= -0.05:
        label = "negative"
    elif compound >= 0.20:
        label = "positive"
    else:
        label = "neutral"

    confidence = min(0.99, abs(compound) * 1.3 + 0.3)
    return label, compound, confidence


def run_sentiment(memory: SharedMemory) -> None:
    start = time.time()
    memory.agent_logs.append(AgentLog("Sentiment Agent", "running", f"Scoring {len(memory.raw_posts)} posts...", time.time()))

    results = []
    for post in memory.raw_posts:
        label, score, confidence = _score(post.title, post.body)
        clean = _clean_text(f"{post.title}. {post.body}")
        results.append(ProcessedPost(
            id=post.id,
            title=post.title,
            body=post.body,
            clean_text=clean,
            score=post.score,
            num_comments=post.num_comments,
            subreddit=post.subreddit,
            author=post.author,
            created_utc=post.created_utc,
            url=post.url,
            source=post.source,
            sentiment_label=label,
            sentiment_score=score,
            sentiment_confidence=confidence,
            location=post.location,
        ))

    memory.processed_posts = results
    neg = sum(1 for p in results if p.sentiment_label == "negative")
    pos = sum(1 for p in results if p.sentiment_label == "positive")
    duration = int((time.time() - start) * 1000)
    memory.agent_logs.append(AgentLog(
        "Sentiment Agent", "done",
        f"Scored {len(results)} posts — {neg} negative, {pos} positive, {len(results)-neg-pos} neutral",
        time.time(), duration
    ))
